### ISSUE TITLE
examine HGNC gene table content and import script

### ISSUE THREAD

**Issue by sheridancbio** [knowledgesystems/pipelines-scrum#462]
Modify [ImportGeneData.java](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/java/org/mskcc/cbio/portal/scripts/ImportGeneData.java) to have it work with new HGNC data source files and ensure that all required table fields are being captured (for fields such as EXONIC LENGTH, ...).

- [x] define new gene data file format, ensuring all existing gene info is captured
- [x] import script updated
- [x] gene table and gene alias table has been populated from HGNC
- [x] dump of tables available for assembling seed database (data only dump)
- [ ] document file format
- [x] microRNA need to be loaded from curated source files
- [x] gene sets need to be reloaded
- [x] Yichao confirming with JJ HGNC’s location is hg38 but for locations marked with extra text they are hg19
- [x] Yichao confirming with JJ that for cytoband values like `Xp11.23 not on reference assembly` we trim the extra text

Notes:
 - NCBI Gene info found here (Homo Sapien): https://ftp.ncbi.nlm.nih.gov/gene/DATA/GENE_INFO/Mammalia/
- We have the freedom to define a new gene data file format.
- It would be nice to avoid having to go to multiple sources to complete the new gene data file (gene length from NCBI, everything else from HGNC).  If unavoidable, curation team should write script to pull data together into file that Manda has defined.

**Comment by mandawilson** [comment]
1. Import new genes:

First follow instructions here: https://docs.cbioportal.org/3.-cbioportal-maintenance/updating-gene-and-gene_alias-tables

```
wc -l /data/manda/hgnc-tools/data/gene-import-input-Mar-11-2021.txt
40887

mysql> select count(*) from gene;
+----------+
| count(*) |
+----------+
|    40865 |
+----------+
1 row in set (0.02 sec)

mysql> select count(*) from gene_alias;
+----------+
| count(*) |
+----------+
|    52471 |
+----------+
1 row in set (0.02 sec)

```
Then more steps from the website...

2. Import micrornas (See https://docs.cbioportal.org/2.1.2-deploy-without-docker/import-the-seed-database):

```
 java -cp scripts/target/scripts-*.jar org.mskcc.cbio.portal.scripts.ImportGeneData -microrna /data/manda/hgnc-tools/data/micrornas.tsv 

mysql> select count(*) from gene;
+----------+
| count(*) |
+----------+
|    44790 |
+----------+
1 row in set (0.01 sec)

mysql> select count(*) from gene_alias;
+----------+
| count(*) |
+----------+
|    72517 |
+----------+
1 row in set (0.03 sec)
```

3. Import gene sets:
```
cd core/src/main/scripts; ./importGenesetData.pl --data /data/manda/hgnc-tools/data/msigdb.v6.1.entrez.gmt --new-version msigdb_6.1

mysql> select count(*) from geneset;
+----------+
| count(*) |
+----------+
|    17786 |
+----------+
1 row in set (0.00 sec)
```

4. Do dump.
```
mysqldump -u root -p --no-create-info cgds_hgnc_debug | gzip -9 -c > cgds_hgnc_debug.hgnc_genes_GRCh37_Mar.sql.gz
```
Count number of gene sets we had in the seed db? Reimported updated seed db to check.

```
mysql> select count(*) from geneset;
+----------+
| count(*) |
+----------+
|    17786 |
+----------+
1 row in set (0.00 sec)
```

**Comment by mandawilson** [comment]
Creating the website and importer:

I imported one study into the cgds_hgnc_debug  database on dashi-dev : http://dashi-dev.cbio.mskcc.org:8080/hgnc-debug

To import (on dashi-dev): /data/portal-cron/scripts/import-hgnc.sh

First you have to make sure whatever study you want is here: /data/portal-cron/cbio-portal-data/datahub/

Using this method `git lfs pull -I public/brca_tcga`, explained here: https://github.com/cBioPortal/datahub

**Comment by mandawilson** [comment]
Curators do not have permission to update the data hub repo (they have to run as cbioportal_impoter).

To set up the entire data hub repo would probably take up too much space (> 100G?).

Maybe run on pipelines instead?  Just use the data hub repo there that is already getting updated nightly.

- [ ] If we run on pipelines, the `buildhgncproductionjar.sh` and possibly other `build*.sh` files need updating -- the comments at the end about manually copying the file needs to be updated to the new location/filename.

- [ ] If we do not run on pipelines undo my last commit here: `dashi-dev:/data/manda/pipelines-configuration/properties/import-hgnc`

**Comment by sheridancbio** [comment]
Curators have followed up on their needs. They have requested the ability to import modified versions of the datahub public studies without the requirement to commit the modifications into the datahub repository. (Or, at the least, without changing the head of the master branch. We contemplated using a separate branch for the hgnc versions of the public studies and having the importer switch the branch of the clone before importing into the hgnc database and then return the branch to the master branch after import was complete, but this level of manipulation of the repository and clone seemed too error-prone.)

The planned approach was to return to using dashi-dev as the import system. The path /data/portal-cron/cbio-portal-data/datahub has been purged of the github repository elements (the .git subdirectory mainly) and all study folders under the "public" top level directory have been rsynced to hold the checked out datahub data from the pipelines server (not the prior object reference hash placeholder files). Directory permissions were set to allow files and directories to be modified by anyone in the unix group "schultz", and the curators plan to process the datahub studies through the gene reference rewriting script and then copy over the updated study data into these directories. Then a dashi-dev importer jar and import script can be run on the dashi-dev server which will import marked studies (hgnc-portal column) into the dashi-dev database.

The execution of this import script will be enabled through either authorities to allow any user to run the import script directly, or by touching a trigger file and having a frequently running crontab check for the trigger file and performing a reset of the trigger followed by an execution of the import script.

**Comment by sheridancbio** [comment]
The script "import-hgnc.sh" on pipelines seems to be the most recent edit. I have compared this script to the one on dashi-dev at path: /data/portal-cron/scripts/import-hgnc.sh
There are two latest changes:
- the importer jar was given a more specific filename (from cbioportal-importer.jar to hgnc-importer.jar)
- the section which performed the datahub fetch has been commented out.

I have added these changes to the dashi-dev copy of the script (changing the commenting out to a conditional if) and also:
-  added a trigger-file launch mechanic so that import will run as cbioportal_importer
- created a utility program (/data/portal-cron/scripts/request-hgnc-import) which will create the trigger file
- simplified the script - removing umask and other bits which were there to try to support an arbitrary user execution mode
- added a fast check so that the script exits if there was no request to import
- added flock protection block to prevent starting the import if the import is currently running

I also renamed the script from "import-hgnc.sh" to "import-hgnc-data.sh"

This script will run on a frequent crontab entry.
